### DISCUSSION TITLE
Pie menus for curve types in the editor

### DISCUSSION THREAD

**Discussion by minetoblend** [ppy/osu#24482] [Ideas]
I think pie menus would be a nice improvement over the dropdown menu to change the curve type of slider control points.

I made a prototype of this in the osucad editor (video below) and had a few mappers try it at coe, and people seemed to like it a lot.

https://github.com/ppy/osu/assets/8039761/ec9e8719-de33-49f3-82ea-6a691de2a219

One of the main advantages is that they can be used in a single drag gesture (you can select an option by just releasing the mouse at a position in the general direction of the menu option) instead of having to use 2 clicks (currently 3 even, since it's hidden in a nested menu). With a tiny bit of muscle memory it becomes super quick to change between curve types since you just need to remember the general direction of a curve type and then do a drag gesture in that direction. For people who do not like to use drag gestures, it can still be used like a regular menu, by releasing right after clicking, and then selecting a menu item.

I think that the menu should also be triggered by something other than right click, I used ctrl-click on a controlpoint in my own editor, which seems to be unused in lazer as well. The context menu can even stay that way, for people that prefer to use that.

I'm open to implementing this myself if this gets accepted, though I'd likely need a bit of guidance with the osu! framework since I'm not super familiar with it yet. I think this would work somewhat similar to `IHasContextMenu`, so it would require a change to osu!framework as well. 

A couple ideas I had for the details of the menu:
I think the way pie menus are done in blender (picture below) is really good, so I kind of want to make it similar to that.
![image](https://github.com/ppy/osu/assets/8039761/0f15ca42-2ca4-4eb1-a8d4-868bbafc9618)
In blender, the menu items aren't positioned automatically, but each menu item has a fixed spot, so even if one option isn't available, the menu items are always in the same spot. They all have icons too which I think makes for a nice touch. I have experience making icons, I also made all the icons for osucad myself (link below), so I could make an icon for each curve type, matching the style of the other lazer icons.
https://www.figma.com/file/VXj8D0lMxNtmFAuiDiKSUB/osucad-icons

**Comment by peppy** [comment]
I'm not ready to overload even more UX patterns in the editor. This is only really saving one click and only for one specific flow. It will also not work on touch well.

**Reply by minetoblend** [reply]
Fair enough, the touch issues are something that I haven't considered and I agree with that. I do think that the improvement goes beyond just saving a click though since it also greatly increases the area the mouse has to aim for (My aim in osu sadly does not translate very well into aiminig for context menu items). Perhaps this idea could be revisited some time in the future, when other use cases are found where pie menus would be a good fit.
